[WIP] Preliminary docker containers for Travis-CI and Deployment

drbenmorgan opened the pull request:
Fixes #2 

This implements basic Dockerfiles for building linux images plus a Linuxbrew layer with installs of all software needed by Falaise. The base linux systems are

- CentOS6
- Ubuntu16.04
- CentOS7
- Ubuntu14.04

The latter two are WIP at the moment. The CentOS6 and Ubuntu16.04 images are already built and pushed to [Docker Hub](https://hub.docker.com/r/supernemo/), with a PR in progress to integrate these in Travis-CI: https://github.com/SuperNEMO-DBD/Falaise/pull/38

A Dockerfile using the above Ubuntu16.04 image as a base is also supplied which adds an additional layer including an install of Falaise. This is intended for deployment to any compute resources that can support Docker containers.

Further testing and work on tagging is needed, plus automated builds of the images.

drbenmorgan committed:
Initial reorganization of files

Simplify layout by flattening containers/ level to just docker/.
Update README to reflect this, and move CVMFS notes into dedicated
subdirectory. Update .gitignore for new layout

drbenmorgan committed:
Preliminary reorganisation

Flatten containers subdirectory for simplicity. Update READMEs as
needed.

drbenmorgan committed:
Add base Ubuntu 16.04 Dockerfile

Update dockerfile to build Ubuntu 16.04 based image, adding:

- System requirements for Falaise install
- Base brew system
- Bootstrap packages
- All Formulae installed by `brew install falaise --only-dependencies`

This shows similar overall size with differing layers, so can
use as a suitable base image for creating runtime or test images.

TODO: Roll out to Ubuntu 14 and CentOS. Decide on repo/tagging names.

drbenmorgan committed:
Update Centos6 Dockerfile

Provide base system for Falaise following that in Ubuntu 16.

drbenmorgan committed:
Base Dockerfiles for Falaise CI and Production

Update CentOS and Ubuntu Dockerfiles to provide a base install
of the system and linuxbrew requirements for Falaise.

Change USER to linuxbrew and installation of brew to $HOME/.linuxbrew
for future compatibility with upstream linuxbrew.

Use base Ubuntu 16.04 image to create image including and install
of Falaise. Update docker README to describe use and build of
images.

lemiere commented:
Few comments about my experience : 
I did : 
`docker pull supernemo/falaise-ubuntu1604-base`
Falaise is not install because that image is prepared to use Travis-CI.

So I tried to install stuff by myself using apt-get install.
In that purpose, I need to be root (sudoer). It should be added to the user configuration in dockerfile.


I tried to install falaise using brew. I remarked that falaise is downloaded from Warwick server and not github. (outdated reason) 

I have been confront to the known issue [https://github.com/SuperNEMO-DBD/homebrew-cadfael/issues/22](url).

I will wait for update version and try again.

lemiere commented:
Another experience, I did: 
` docker pull supernemo/falaise`
Next step is : 
` docker run -i -t supernemo/falaise`
Falaise is install but I experienced the same issue about sudo opportunity.

But : 
`flsimulate-configure -o toto.conf -t "urn:snemo:demonstrator:simulation:2.1"`
`QXcbConnection: Could not connect to display`
`Aborted (core dumped)`

So I tested that trick : 
`docker run -i -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix -t supernemo/falaise`

I openned the door to the hell......
` flsimulate-configure  \`
` -t "urn:snemo:demonstrator:simulation:2.1"  \`
 ` -o myprofile.conf`
--------------------
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for /tmp/bayeux-20170621-17944-87xjvu/Bayeux-3.0.0/source/bxbrio/include/brio/detail/brio_record.h
   requested to autoload type TArrayCMod
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for /tmp/bayeux-20170621-17944-87xjvu/Bayeux-3.0.0/source/bxbrio/include/brio/detail/brio_record.h
   requested to autoload type brio_record

 *** Break *** segmentation violation
 Generating stack trace...
 0x00007fd70b4afcd6 in QDBusConnectionInterface::isServiceRegistered(QString const&) const + 0x86 from /home/linuxbrew/.linuxbrew/Cellar/qt5-base/5.8.0_1/plugins/platforms/../../lib/libQt5DBus.so.5
 0x00007fd70b5e366b in <unknown> from /home/linuxbrew/.linuxbrew/Cellar/qt5-base/5.8.0_1/plugins/platforms/../../lib/libQt5XcbQpa.so.5
 0x00007fd70b5d73b2 in <unknown> from /home/linuxbrew/.linuxbrew/Cellar/qt5-base/5.8.0_1/plugins/platforms/../../lib/libQt5XcbQpa.so.5
 0x00007fd70b5639ce in QXcbIntegration::accessibility() const + 0x2e from /home/linuxbrew/.linuxbrew/Cellar/qt5-base/5.8.0_1/plugins/platforms/../../lib/libQt5XcbQpa.so.5
 0x00007fd7126cd9ec in QAccessible::isActive() + 0x1c from /home/linuxbrew/.linuxbrew/opt/qt5-base/lib/libQt5Gui.so.5
 0x00007fd7126cdbc5 in QAccessible::updateAccessibility(QAccessibleEvent*) + 0x15 from /home/linuxbrew/.linuxbrew/opt/qt5-base/lib/libQt5Gui.so.5
 0x00007fd712e0c21a in QLabel::setText(QString const&) + 0x18a from /home/linuxbrew/.linuxbrew/opt/qt5-base/lib/libQt5Widgets.so.5
 0x00007fd717dbcbd4 in datatools::configuration::ui::variant_repository_viewer::variant_repository_viewer(QWidget*) + 0xf2 from /home/linuxbrew/.linuxbrew/lib/libBayeux.so.3
 0x00007fd717dbb2f5 in datatools::configuration::ui::variant_repository_dialog::_construct() + 0x7b from /home/linuxbrew/.linuxbrew/lib/libBayeux.so.3
 0x00007fd717dbb7ca in datatools::configuration::ui::variant_repository_dialog::set_repository(datatools::configuration::variant_repository&) + 0xfe from /home/linuxbrew/.linuxbrew/lib/libBayeux.so.3
 0x00007fd717dbb869 in datatools::configuration::ui::variant_repository_dialog::variant_repository_dialog(datatools::configuration::variant_repository&, QWidget*) + 0x49 from /home/linuxbrew/.linuxbrew/lib/libBayeux.so.3
 0x00007fd717d5b047 in datatools::configuration::variant_service::_do_variant_launch_gui_() + 0xd1 from /home/linuxbrew/.linuxbrew/lib/libBayeux.so.3
 0x00007fd717d5b1a5 in datatools::configuration::variant_service::_do_build_() + 0x9d from /home/linuxbrew/.linuxbrew/lib/libBayeux.so.3
 0x00007fd717d5b47a in datatools::configuration::variant_service::configure(datatools::configuration::variant_service::config const&, bool) + 0x250 from /home/linuxbrew/.linuxbrew/lib/libBayeux.so.3
 0x0000000000412cfc in FLSimulateConfig::do_flsimulate_config(int, char**) + 0xa2 from flsimulate-configure
 0x0000000000410738 in main + 0x20 from flsimulate-configure
 0x00007fd716615830 in __libc_start_main + 0xf0 from /lib/x86_64-linux-gnu/libc.so.6
 0x00000000004107c9 in _start + 0x29 from flsimulate-configure
terminate called after throwing an instance of 'std::logic_error'
  what():  [datatools::base_service& datatools::service_manager::grab_service(const string&):163: No service named 'P�▒
Aborted (core dumped)
`

drbenmorgan commented:
On the `apt-get` side, I'll add sudo permission to the `linuxbrew` user.

I'll get  https://github.com/SuperNEMO-DBD/homebrew-cadfael/issues/22 fixed first, along with the bump in the Bayeux/Falaise tarball URLs.

For sharing X11, I hadn't even though about this yet - if you haven't seen these already, a few things I found via a _very_ quick googling:

http://wiki.ros.org/docker/Tutorials/GUI

https://blog.jessfraz.com/post/docker-containers-on-the-desktop/

drbenmorgan committed:
[CentOS6] Install epel-release rpm

Required HEP_OSlibs rpm requires packages from epel repo. Enable
epel repo by installing appropriate rpm as intermediate install step
before main dependencies.

drbenmorgan committed:
[CentOS7] Install falaise dependency layer
